packages/licenses/src/lockfileToLicenseNodeTree.ts

```diff
@@ -0,0 +1,161 @@
+import path from "path";
+import { Lockfile } from "@pnpm/lockfile-types";
+import { nameVerFromPkgSnapshot } from "@pnpm/lockfile-utils";
+import {
+  lockfileWalkerGroupImporterSteps,
+  LockfileWalkerStep,
+} from "@pnpm/lockfile-walker";
+import { DependenciesField, PackageManifest } from "@pnpm/types";
+import * as dp from "dependency-path";
+import { GetPackageInfoFunction } from "./licenses";
+
+export interface LicenseNode {
+  name?: string;
+  version?: string;
+  license: string;
+  licenseContents?: string;
+  packageManifest: PackageManifest;
+  path: string;
+  vendorName?: string;
+  vendorUrl?: string;
+  integrity?: string;
+  requires?: Record<string, string>;
+  dependencies?: { [name: string]: LicenseNode };
+  dev: boolean;
+}
+
+export type LicenseNodeTree = Omit<
+  LicenseNode,
+  "path" | "license" | "vendorName" | "vendorUrl" | "packageManifest"
+> & {
+  install: string[];
+  remove: string[];
+  metadata: Object;
+};
+
+export interface LicenseExtractOptions {
+  virtualStoreDir: string;
+  modulesDir?: string;
+  dir: string;
+  getPackageInfo: GetPackageInfoFunction;
+}
+
+/**
+ * private
+ */
+export async function lockfileToLicenseNode(
+  step: LockfileWalkerStep,
+  options: LicenseExtractOptions
+) {
+  const dependencies = {};
+  for (const dependency of step.dependencies) {
+    const { depPath, pkgSnapshot, next } = dependency;
+    const { name, version } = nameVerFromPkgSnapshot(depPath, pkgSnapshot);
+
+    const modules = path.join(
+      options.dir,
+      options.virtualStoreDir,
+      dp.depPathToFilename(depPath),
+      options.modulesDir ?? "node_modules"
+    );
+
+    let packageDetails;
+    try {
+      packageDetails = await options.getPackageInfo({
+        name,
+        version,
+        prefix: modules,
+      });
+    } catch (err: unknown) {}
```

Comment:
I have the problem that sometimes an optional dependency is listed in the lock file but not stored in modules directory and then the fetching fails. In that case I am skipping/ignoring this package so it won't get listed.

Reply:
No, it is not OK to skip such issues. It would mean that your license checker doesn't check all the dependencies, so a package with a bad license might pass under your radar. Besides, now that you read from the CAFS, the optional deps will always be found. So this try/catch is not needed.

Reply:
I think this problem is resolved now. I have removed it, while making this change I did struggle to make it work with PNPM repo and the `pkg`-package as it's a `tarball` and the workspace projects but I think I cracked it. Only definitely not 100% sure it's the right approach